### ISSUE TITLE
Three-Character Extension Note Is Incorrect For Directory.EnumerateFiles Method

### ISSUE THREAD

**Issue by gewarren** [dotnet/dotnet-api-docs#4052]
_From @RedQuark on Sunday, March 22, 2020 12:58:32 AM_

**Link to article:**
https://docs.microsoft.com/en-us/dotnet/api/system.io.directory.enumeratefiles?view=netcore-3.1

**Problem:**
The article states that with a search for files ending in three-character extensions using an asterisk wildcard (e.g. **"*.xls"**) "the method returns files with extensions that _**begin**_ with the specified extension." The example says that the above search term will return both "books.xls" and "books.xlsx".

Although the command `dir ".xls"` does indeed list both files in Windows 10's command prompt, the method `Directory.EnumerateFiles(".", "*.xls") `only returns "books.xls" -- at least when running .NET Core 3.1.2 on WIndows 10 version 1909.

_Copied from original issue: MicrosoftDocs/feedback#2604_

**Comment by gewarren** [comment]
_From @welcome[bot] on Sunday, March 22, 2020 12:58:33 AM_

Thank you for creating the issue! One of our team members will get back to you shortly with additional information. If this is a product issue, please close this and contact the particular product's support instead (see https://support.microsoft.com/allproducts for the list of support websites).

**Comment by gewarren** [comment]
_From @gewarren on Thursday, March 26, 2020 10:42:39 PM_

@carlossanlop Can you take a look at this feedback?

**Comment by gewarren** [comment]
_From @carlossanlop on Thursday, March 26, 2020 11:06:01 PM_

Thanks for bringing it to my attention, @gewarren.

The documentation is incorrect for this code example.

`EnumerateFiles` will only return files that have the `xls` extension if the `searchPattern` argument is set to `*.xls`.

To enumerate files that end both with `xls` and with `xlsx`, the `searchPattern` argument should be `*.xls*`:

```cs
Directory.EnumerateFiles(".", "*.xls*"))
```

**Comment by Roald87** [comment]
> `EnumerateFiles` will only return files that have the `xls` extension if the `searchPattern` argument is set to `*.xls`.

This is not what the [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.io.directory.enumeratefiles?view=netframework-4.6.1) says:

> When you use the asterisk wildcard character in a `searchPattern` such as "*.txt", the number of characters in the specified extension affects the search as follows:
> - If the specified extension is exactly three characters long, the method returns files with extensions that begin with the specified extension. For example, "*.xls" returns both "book.xls" and "book.xlsx".

Finally the behavior is also different between .NET Core and .NET Framework. If I run the same code with .NET Core 2.1 as with .NET Framework 4.6.1, then Core code will only find the exact extension, whereas Framework will also find the others such as `.xlsx`.

**Comment by jwdonahue** [comment]
I was so shocked when I read that doc, that I wrote a .NET Core 3.1 console app to test it, and was unable to duplicate the stated behavior. After finding this issue here, I targeted .NET 5.0 and was also unable to replicated that stated behavior. I find it incomprehensible that anyone would design a feature to behave as the doc specifies.  This documentation cost me several hours or development time this afternoon, when I added code to my project to account for it, and then couldn't find any way to drive coverage into that new code from my unit test. I finally gave up and decided the documentation was bogus.

I have always disliked the Exploder search behavior, which behaves exactly like this doc describes (perhaps that's an earlier .NET framework behavior?). I found it shocking the first time I witnessed it.  Cmd.exe's `dir` command does not behave that way.  A glob search consists of literal content, '*' or '?', where the literal content must match exactly and the globs have their respective meanings, period. No goofy "if the extension faces east and you close one eye..." behaviors.

I'll be happy to issue a PR with the content expunged, but if the doc requires special notes for different versions, I want no part of it.

**Comment by gewarren** [comment]
I have a PR to update the remarks for each overload that takes a search pattern. How does this look?

![image](https://user-images.githubusercontent.com/24882762/106046348-febb9800-6096-11eb-8922-c84f2bd183bf.png)

**Comment by Roald87** [comment]
Maybe also add a table with examples for both two and three character extensions for both .NETs. Then it is immediately clear that .NET Framework works not as expected.

For example:

| files               | searchPattern | .NET Core returns | .NET Framework returns |
|---------------------|---------------|-------------------|------------------------|
| file.ai, file.aif   | *.ai          | file.ai           | file.ai                |
| book.xls, book.xlsx | *.xls         | book.xls          | book.xls, book.xlsx    |
| file.ai, file.aif   | ?.ai          | file.ai           | file.ai                |
| book.xls, book.xlsx | ?.xls         | book.xls          | book.xls               |

Best would be of course to change what it returns. Needing such a long note for such a simple function is a warning sign I would say.

**Comment by gewarren** [comment]
@Roald87 Thanks, I incorporated the table. I think it's pretty unlikely that .NET Framework code will get changed at this point.

**Comment by ZonderP** [comment]
The '?.xls' example doesn't work for me.
And in my understanding of the wildcard characters, it also isn't supposed to work.
Thus I think the documentation is wrong in this regards.
Actually I came across this issue by an even much more annoying problem.
I have project creating a DLL targeting .netstandard 2.0 which implements a function using EnumerateFiles.
So far I only consumed it by .Net Core projects and the function in the consumed DLL which uses EnumerateFiles with '*.dll' as SearchPattern worked fine.
Now I created a .NET Framework 4.6.1 project where I also consume that function, but then - although the consumed DLL targets .netstandard 2.0 - I nonetheless get the behavior which is described in the Note '.NET Framework only: ....', namely that also files ending with e.g. '.dllX' get returned.
